spdlog/src/formatter/pattern_formatter/runtime.rs

```diff
@@ -0,0 +1,270 @@
+/// # assert_eq!(
+/// #     sink.clone_string().replace("\r", ""),
+/// /* Output */ "[info] Interesting log message\n"
+/// # );
+/// # Ok(()) }
+/// ```
+/// 
+/// ## With Custom Patterns
+/// 
+/// ```
+/// use std::fmt::Write;
+///
+/// use spdlog::{
+///     formatter::{pattern, Pattern, PatternContext, PatternFormatter, runtime_pattern, RuntimePattern},
+///     Record, StringBuf, info
+/// };
+/// 
+/// #[derive(Default, Clone)]
+/// struct MyPattern;
+/// 
+/// impl Pattern for MyPattern {
+///    fn format(&self, record: &Record, dest: &mut StringBuf, _: &mut PatternContext) -> spdlog::Result<()> {
+///        write!(dest, "My own pattern").map_err(spdlog::Error::FormatRecord)
+///    }
+/// }
+///
+#[doc = include_str!(concat!(env!("OUT_DIR"), "/test_utils/common_for_doc_test.rs"))]
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let template = "[{level}] {payload} - {$mypat1} {$mypat2}{eol}";
+/// # // TODO: Directly pass the closure to runtime_pattern! macro
+/// fn pat() -> impl Pattern { pattern!("[{level_short}-{level}]") }
+/// let formatter = PatternFormatter::new(
+///     runtime_pattern!(
+///         template,
+///         {$mypat1} => MyPattern::default,
+///         {$mypat2} => pat
+///     )?
+/// );
+/// # let (doctest, sink) = test_utils::echo_logger_from_formatter(
+/// #     Box::new(formatter),
+/// #     None
+/// # );
+///
+/// info!(logger: doctest, "Interesting log message");
+/// # assert_eq!(
+/// #     sink.clone_string().replace("\r", ""),
+/// /* Output */ "[info] Interesting log message - My own pattern [I-info]\n"
+/// # );
+/// # Ok(()) }
+/// ```
+/// 
+/// [`pattern!`]: crate::formatter::pattern
+#[derive(Clone)]
+pub struct RuntimePattern(Patterns);
+
+impl RuntimePattern {
+    // Private function, do not use in your code directly.
+    #[doc(hidden)]
+    pub fn __with_custom_patterns(template: &str, registry: PatternRegistry) -> Result<Self> {
+        Template::parse(template)
+            .and_then(|template| {
+                Synthesiser::new(registry)
+                    .synthesize(template)
+                    .map(RuntimePattern)
+            })
+            .map_err(|err| Error::BuildPattern(BuildPatternError(err)))
+    }
+}
+
+impl Pattern for RuntimePattern {
+    fn format(
+        &self,
+        record: &Record,
+        dest: &mut StringBuf,
+        ctx: &mut PatternContext,
+    ) -> Result<()> {
+        for pattern in &self.0 {
+            pattern.format(record, dest, ctx)?;
+        }
+        Ok(())
+    }
+}
+
+struct Synthesiser {
+    registry: PatternRegistry,
+}
+
+impl Synthesiser {
+    fn new(registry: PatternRegistry) -> Self {
+        Self { registry }
+    }
+
+    fn synthesize(&self, template: Template) -> PatternParserResult<Patterns> {
+        self.build_patterns(template, false)
+    }
+
+    fn build_patterns(
+        &self,
+        template: Template,
+        mut style_range_seen: bool,
+    ) -> PatternParserResult<Patterns> {
```

Comment:
`build_patterns` will recursively call itself with an `style_range_seen = true` argument, while `synthesis` calls it initially with `style_range_seen = false`.

Keeping this way we can hide the initial value of this argument instead of the caller being responsible for specifying it.